File: C++/프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.cpp
#include <string>
#include <vector>

using namespace std;

bool check_limit(vector<int> &diffs, vector<int> &times, int &n, long long &limit, int level) {
    long long total_time = 0;
    int time_prev = 0;
    for (int i = 0; i < n; i++) {
        int diff = diffs[i];
        int time_cur = times[i];
        if (i > 0) time_prev = times[i-1];
        
        if (diff > level) total_time += (diff - level) * (time_cur + time_prev) + time_cur;
        else total_time += time_cur;
        
   }
    if (total_time > limit) return false;
    return true;
}

int solution(vector<int> diffs, vector<int> times, long long limit) {
    int answer = 0;
    int n = diffs.size();
    int s = 1, e = 100000;
    while (s <= e) {
        int mid = (s + e) / 2;
        if (check_limit(diffs, times, n, limit, mid)) {
            answer = mid;
            e = mid - 1;
        }
        else {
            s = mid + 1;
        }
    }
    return answer;
}

File: Python3/프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def check(level, diffs, times, limit):
    total_time = 0
    time_prev = 0
    for i in range(len(diffs)):
        diff = diffs[i]
        time_cur = times[i]
        if diff <= level:
            total_time += time_cur
        else:
            total_time += (time_cur + time_prev) * (diff - level) + time_cur
        time_prev = time_cur
    
    if total_time <= limit:
        return True
    return False
    
        
def solution(diffs, times, limit):
    answer = 0
    s, e = 1, 100000
    while s <= e:
        mid = (s + e) // 2
        if check(mid, diffs, times, limit):
            answer = mid
            e = mid - 1
        else:
            s = mid + 1
    return answer
